File: netlogo-gui/src/main/sdm/gui/AggregateModelEditor.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.sdm.gui;

import org.jhotdraw.framework.DrawingEditor;
import org.jhotdraw.framework.DrawingView;
import org.jhotdraw.framework.Figure;
import org.jhotdraw.framework.FigureEnumeration;
import org.jhotdraw.framework.Tool;
import org.jhotdraw.framework.ViewChangeListener;
import org.jhotdraw.util.UndoManager;
import org.nlogo.core.CompilerException;
import org.nlogo.api.CompilerServices;
import org.nlogo.core.I18N;
import org.nlogo.core.TokenType;
import org.nlogo.sdm.Model;
import org.nlogo.window.EditDialogFactoryInterface;

import javax.swing.JMenuBar;
import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Dimension;

strictfp class AggregateModelEditor
    extends javax.swing.JFrame
    implements
    DrawingEditor,
    org.nlogo.window.Event.LinkChild {

  /// Constants

  private static final Dimension WINDOW_SIZE = new Dimension(700, 550);
  private static final Dimension VIEW_SIZE = new Dimension(800, 1000);

  /// Instance variables

  private final org.nlogo.editor.Colorizer colorizer;
  private final java.awt.Component linkParent;
  private final org.nlogo.window.MenuBarFactory menuBarFactory;
  private final CompilerServices compiler;
  private final AggregateDrawing drawing;
  private AggregateModelEditorToolBar toolbar;
  private final UndoManager undoManager;
  private Tool currentTool;
  private Tool selectionTool;
  private DrawingView view;
  private AggregateTabs tabs;
  private org.nlogo.window.ErrorLabel errorLabel;
  private final EditDialogFactoryInterface dialogFactory;

  AggregateModelEditor
      (java.awt.Component linkParent,
       org.nlogo.editor.Colorizer colorizer,
       org.nlogo.window.MenuBarFactory menuBarFactory,
       CompilerServices compiler,
       EditDialogFactoryInterface dialogFactory) {
    this(linkParent, colorizer, menuBarFactory, null, compiler, dialogFactory);
  }

  AggregateModelEditor
      (java.awt.Component linkParent,
       org.nlogo.editor.Colorizer colorizer,
       org.nlogo.window.MenuBarFactory menuBarFactory,
       AggregateDrawing drawing,
       CompilerServices compiler,
       EditDialogFactoryInterface dialogFactory) {
    super(I18N.guiJ().get("menu.tools.systemDynamicsModeler"), linkParent.getGraphicsConfiguration());
    undoManager = new UndoManager();

    this.linkParent = linkParent;
    this.colorizer = colorizer;
    this.menuBarFactory = menuBarFactory;
    this.compiler = compiler;
    this.dialogFactory = dialogFactory;

    Wrapper.reset();

    // You might think it'd make more sense to dispose of the window --
    // not hide it -- when it's closed. But disposing and then recreating
    // the window causes the OS X screen menu bar to go nuts, for
    // reasons I don't pretend to understand. So, the editor gets made
    // at most once per model. After it's been made, it hides and shows
    // but is not disposed of until the model is closed. - AZS 6/18/05
    setDefaultCloseOperation(HIDE_ON_CLOSE);

    this.drawing = (drawing == null) ? new AggregateDrawing() : drawing;

    DrawingView drawingView =
        new AggregateDrawingView
            (this, VIEW_SIZE.width, VIEW_SIZE.height);
    drawingView.setDrawing(this.drawing);


    open(drawingView);

    pack();
    setVisible(true);
  }

  void setError(org.nlogo.api.SourceOwner owner, CompilerException e) {
    errorLabel.setError(e, owner.headerSource().length());
    tabs.setError(e);
  }

  public Object getLinkParent() {
    return linkParent;
  }

  void setToolbar(AggregateModelEditorToolBar toolbar) {
    this.toolbar = toolbar;
  }

  void inspectFigure(org.jhotdraw.framework.Figure f) {
    try {
      org.nlogo.api.Editable target =
          (org.nlogo.api.Editable) f;

      // makes a dialog and returns a boolean result. we ignore the result - ST 3/2/09
      dialogFactory.canceled(this, target);

      if (f instanceof ModelElementFigure && ((ModelElementFigure) f).dirty()) {
        new org.nlogo.window.Events.CompileAllEvent().raise(this);
        new org.nlogo.window.Events.DirtyEvent().raise(this);
      }

      f.invalidate();
    } catch (ClassCastException ex) {
      // if it's not editable, do nothing
      org.nlogo.api.Exceptions.ignore(ex);
    }
  }

  public AggregateDrawing getDrawing() {
    return drawing;
  }

  /**
   * Translates the model into NetLogo code.
   */
  public String toNetLogoCode() {
    String src = "";

    if (drawing != null) {
      org.nlogo.sdm.Translator translator =
          new org.nlogo.sdm.Translator(drawing.getModel(), compiler);
      src = translator.source();
    }

    return src;
  }

  public void setTool(Tool t) {
    currentTool = t;
    if (tool() != null) {
      tool().activate();
    }
  }

  public void setSelectionTool() {
    if (selectionTool != null) {
      setTool(selectionTool);
    }
  }

  private void open(final DrawingView view) {
    getContentPane().setLayout(new java.awt.BorderLayout());
    this.view = view;

    errorLabel = new org.nlogo.window.ErrorLabel();
    getContentPane().add(errorLabel, BorderLayout.NORTH);

    AggregateEditorTab editorTab =
        new AggregateEditorTab(this, new AggregateModelEditorToolBar(this, drawing.getModel()), (Component) view);
    tabs =
        new AggregateTabs(editorTab, this, colorizer);
    getContentPane().add(tabs, BorderLayout.CENTER);


    // Build the menu bar. For OS X, we add a bunch of the menus from app
    // so that the screen menu bar looks consistent. - AZS 6/17/05
    JMenuBar menuBar = new javax.swing.JMenuBar();

    boolean isOSX = System.getProperty("os.name").startsWith("Mac");

    if (isOSX) {
      menuBar.add(menuBarFactory.createFileMenu());
    }

    org.jhotdraw.util.CommandMenu editMenu =
        new org.jhotdraw.util.CommandMenu(I18N.guiJ().get("menu.edit"));
    editMenu.add(new org.jhotdraw.util.UndoCommand(I18N.guiJ().get("menu.edit.undo"), this));
    editMenu.add(new org.jhotdraw.util.RedoCommand(I18N.guiJ().get("menu.edit.redo"), this));
    menuBar.add(editMenu);

    if (isOSX) {
      menuBar.add(menuBarFactory.createToolsMenu());
      javax.swing.JMenuItem zoomMenu =
          menuBar.add(menuBarFactory.createZoomMenu());
      zoomMenu.setEnabled(false);
      menuBar.add(zoomMenu);
    }

    menuBar.add(new org.nlogo.swing.TabsMenu(I18N.guiJ().get("menu.tabs"), tabs));

    if (isOSX) {
      menuBarFactory.addHelpMenu(menuBar);
    }

    setJMenuBar(menuBar);

    setSize(WINDOW_SIZE);
    setVisible(true);

    selectionTool = new InspectionTool(this);
    setTool(selectionTool);

    setVisible(true);

    org.nlogo.awt.EventQueue.invokeLater
        (new Runnable() {
          public void run() {
            toFront();
          }
        });
  }


  /// From interface DrawingEditor

  /**
   * Retrieve the active view from the window
   * Gets the current drawing view.
   *
   * @see DrawingEditor
   */
  public DrawingView view() {
    return view;
  }

  public DrawingView[] views() {
    return new DrawingView[]{view()};
  }

  /**
   * Gets the current tool.
   *
   * @see DrawingEditor
   */
  public Tool tool() {
    return currentTool;
  }

  public void toolDone() {
    setSelectionTool();
    toolbar.popButtons();
  }

  public UndoManager getUndoManager() {
    return undoManager;
  }

  /**
   * Empty implementation.
   *
   * @see DrawingEditor
   */
  public void figureSelectionChanged(DrawingView view) {
  }

  /**
   * Empty implementation.
   *
   * @see DrawingEditor
   */
  public void showStatus(String str) {
  }

  /**
   * Register to hear when the active view is changed.  For Single document
   * interface, this will happen when a new drawing is created.
   *
   * @see DrawingEditor
   */
  public void addViewChangeListener(ViewChangeListener vcl) {
  }

  /**
   * Remove listener
   *
   * @see DrawingEditor
   */
  public void removeViewChangeListener(ViewChangeListener vcl) {
  }


}

File: netlogo-gui/src/main/sdm/gui/AggregateModelEditor.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.sdm.gui

import java.awt.{ Component, Dimension, Graphics }
import java.awt.event.ActionEvent
import javax.swing.{ AbstractAction, JFrame, JMenuBar, JMenuItem, JPopupMenu, WindowConstants },
  WindowConstants.HIDE_ON_CLOSE
import javax.swing.border.LineBorder
import javax.swing.plaf.basic.BasicMenuUI

import org.jhotdraw.framework.{ DrawingEditor, DrawingView, Figure, Tool, ViewChangeListener }
import org.jhotdraw.util.{ Command, CommandMenu, RedoCommand, UndoCommand, UndoManager }

import org.nlogo.api.{ CompilerServices, SourceOwner }
import org.nlogo.awt.EventQueue
import org.nlogo.core.{ CompilerException, I18N, LiteralParser }
import org.nlogo.editor.Colorizer
import org.nlogo.sdm.Translator
import org.nlogo.swing.{ MenuItem, NetLogoIcon, Utils => SwingUtils }
import org.nlogo.theme.{ InterfaceColors, ThemeSync }
import org.nlogo.window.{ Editable, EditDialogFactory, Events, MenuBarFactory }
import org.nlogo.window.Event.LinkChild

object AggregateModelEditor {
  /// Constants
  val WindowSize: Dimension = new Dimension(700, 550)
  private val ViewSize: Dimension = new Dimension(800, 1000)
}

import AggregateModelEditor._

class AggregateModelEditor(
  linkParent: Component,
  colorizer: Colorizer,
  menuBarFactory: MenuBarFactory,
  val drawing: AggregateDrawing,
  compiler: LiteralParser,
  dialogFactory: EditDialogFactory) extends JFrame(
    I18N.gui.get("menu.tools.systemDynamicsModeler"), linkParent.getGraphicsConfiguration)
  with DrawingEditor
  with LinkChild
  with Events.LoadBeginEvent.Handler
  with ThemeSync
  with NetLogoIcon {

  def this(
    linkParent: Component,
    colorizer: Colorizer,
    menuBarFactory: MenuBarFactory,
    compiler: CompilerServices,
    dialogFactory: EditDialogFactory) =
      this(linkParent, colorizer, menuBarFactory, new AggregateDrawing(), compiler, dialogFactory)

  private val undoManager: UndoManager = new UndoManager()
  private var currentTool: Option[Tool] = None

  locally {
    Wrapper.reset()

    // You might think it'd make more sense to dispose of the window --
    // not hide it -- when it's closed. But disposing and then recreating
    // the window causes the OS X screen menu bar to go nuts, for
    // reasons I don't pretend to understand. So, the editor gets made
    // at most once per model. After it's been made, it hides and shows
    // but is not disposed of until the model is closed. - AZS 6/18/05
    setDefaultCloseOperation(HIDE_ON_CLOSE)
  }

  val view = new AggregateDrawingView(this, ViewSize.width, ViewSize.height)

  view.setDrawing(drawing)

  private val toolbar: AggregateModelEditorToolBar = new AggregateModelEditorToolBar(this, drawing.getModel)

  val tabs: AggregateTabs = {
    val editorTab = new AggregateEditorTab(toolbar, view.asInstanceOf[Component])
    editorTab.setBorder(null)
    val proceduresTab = new AggregateProceduresTab(colorizer)
    new AggregateTabs(this, editorTab, proceduresTab)
  }

  getContentPane.add(tabs)

  private val selectionTool = new InspectionTool(this, drawing.getModel)

  private val menuBar: MenuBar = new MenuBar

  locally {
    // Build the menu bar. For OS X, we add a bunch of the menus from app
    // so that the screen menu bar looks consistent. - AZS 6/17/05

    val isOSX = System.getProperty("os.name").startsWith("Mac");

    if (isOSX) {
      menuBar.add(menuBarFactory.createFileMenu);
    }

    val editMenu = new SyncedCommandMenu(I18N.gui.get("menu.edit"))

    editMenu.add(new UndoCommand(I18N.gui.get("menu.edit.undo"), this))
    editMenu.add(new RedoCommand(I18N.gui.get("menu.edit.redo"), this))

    menuBar.add(editMenu)

    if (isOSX) {
      menuBar.add(menuBarFactory.createToolsMenu)
      val zoomMenu: JMenuItem =
        menuBar.add(menuBarFactory.createZoomMenu)
      zoomMenu.setEnabled(false)
      menuBar.add(zoomMenu)
    }

    menuBar.add(new TabsMenu(I18N.gui.get("menu.tabs"), tabs))

    if (isOSX) {
      val helpMenu = menuBarFactory.createHelpMenu
      menuBar.add(helpMenu)
      try {
        menuBar.setHelpMenu(helpMenu)
      } catch {
        case e: Error => org.nlogo.api.Exceptions.ignore(e)
      }
    }

    setJMenuBar(menuBar)

    setTool(selectionTool)
    setPreferredSize(WindowSize)
    pack()
    setVisible(true)

    EventQueue.invokeLater(new Runnable() {
      def run(): Unit = { toFront() }
    })
  }

  syncTheme()

  def clearError(): Unit = {
    tabs.clearError()
  }

  def setError(owner: SourceOwner, e: CompilerException): Unit = {
    if (e == null)
      tabs.clearError()
    else
      tabs.setError(e, owner.headerSource.length)
  }

  def getLinkParent: Component = linkParent

  def inspectFigure(f: Figure): Unit = {
    f match {
      case target: Editable =>
        // makes a dialog and returns a boolean result. we ignore the result - ST 3/2/09
        dialogFactory.canceled(this, target, false)

        f match {
          case mef: ModelElementFigure if mef.dirty =>
            new org.nlogo.window.Events.CompileAllEvent().raise(this)
            new org.nlogo.window.Events.DirtyEvent(None).raise(this)
          case _ =>
        }

        f.invalidate()
      case _ => // if it's not editable, do nothing
    }
  }

  /**
   * Translates the model into NetLogo code.
   */
  def toNetLogoCode: String = {
    drawing.synchronizeModel()
    new Translator(drawing.getModel, compiler).source
  }

  def setTool(t: Tool): Unit = {
    currentTool = Option(t)
    currentTool.foreach(_.activate())
  }

  def setSelectionTool(): Unit = {
    setTool(selectionTool)
  }


  /// From interface DrawingEditor
  def views: Array[DrawingView] = Array[DrawingView](view)

  /**
   * Gets the current tool.
   *
   * @see DrawingEditor
   */
  def tool: Tool = currentTool.orNull

  def toolDone(): Unit = {
    setSelectionTool()
    toolbar.popButtons()
  }

  def getUndoManager: UndoManager = undoManager

  /**
   * Empty implementation.
   *
   * @see DrawingEditor
   */
  def figureSelectionChanged(view: DrawingView): Unit = { }

  /**
   * Empty implementation.
   *
   * @see DrawingEditor
   */
  def showStatus(str: String): Unit = { }

  /**
   * Register to hear when the active view is changed.  For Single document
   * interface, this will happen when a new drawing is created.
   *
   * @see DrawingEditor
   */
  def addViewChangeListener(vcl: ViewChangeListener): Unit = { }

  /**
   * Remove listener
   *
   * @see DrawingEditor
   */
  def removeViewChangeListener(vcl: ViewChangeListener): Unit = { }

  def handle(e: Events.LoadBeginEvent) {
    undoManager.clearUndos()
    undoManager.clearRedos()
  }

  override def syncTheme(): Unit = {
    menuBar.syncTheme()
    tabs.syncTheme()
    toolbar.syncTheme()
    view.syncTheme()
  }

  private class MenuBar extends JMenuBar with ThemeSync {
    override def paintComponent(g: Graphics) {
      val g2d = SwingUtils.initGraphics2D(g)

      g2d.setColor(InterfaceColors.menuBackground())
      g2d.fillRect(0, 0, getWidth, getHeight)
    }

    override def paintBorder(g: Graphics) {
      val g2d = SwingUtils.initGraphics2D(g)

      g2d.setColor(InterfaceColors.menuBarBorder())
      g2d.drawLine(0, getHeight - 1, getWidth, getHeight - 1)
    }

    def syncTheme {
      getComponents.foreach(_ match {
        case ts: ThemeSync => ts.syncTheme()
        case _ =>
      })
    }
  }

  private class SyncedCommandMenu(name: String) extends CommandMenu(name) with ThemeSync {
    private val menuUI = new BasicMenuUI with ThemeSync {
      override def syncTheme(): Unit = {
        setForeground(InterfaceColors.toolbarText())

        selectionBackground = InterfaceColors.menuBackgroundHover()
        selectionForeground = InterfaceColors.menuTextHover()
        acceleratorForeground = InterfaceColors.toolbarText()
        acceleratorSelectionForeground = InterfaceColors.menuTextHover()
        disabledForeground = InterfaceColors.menuTextDisabled()
      }
    }

    setUI(menuUI)
    syncTheme()

    override def addMenuItem(command: Command, menuItem: JMenuItem) {
      super.addMenuItem(command, new MenuItem(new AbstractAction(command.name) {
        def actionPerformed(e: ActionEvent) {
          command.execute()
        }
      }))
    }

    override def getPopupMenu: JPopupMenu = {
      val menu = super.getPopupMenu

      menu.setBackground(InterfaceColors.menuBackground())
      menu.setBorder(new LineBorder(InterfaceColors.menuBorder()))

      menu
    }

    override def syncTheme(): Unit = {
      menuUI.syncTheme()

      getMenuComponents.foreach(_ match {
        case ts: ThemeSync => ts.syncTheme()
        case _ =>
      })
    }
  }
}
